### PR TITLE
Allow empty strings in config flow form values

### PR DESCRIPTION
**Pull request by eifinger**
### What?

Allow empty strings as values in config flow step forms.

### Why?

The Waze Travel Time integration allows to filter the routes it supplies by strings which have to be part of the route or must not be part of the route. This is done via two string fields in the OptionsFlow.
We have an open issue https://github.com/home-assistant/core/issues/86946 that once set, these fields cannot be unset anymore.

Empty strings are filtered out and not included in the data structure send to the backend.
There the schema is applied to the data here https://github.com/home-assistant/core/blob/dev/homeassistant/data_entry_flow.py#L245

Because the fields are not part of the data received the default value is set for those fields. And this is the value of the field as it was before the OptionsFlow started.
The intention "empty the string again" is ignored.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion: https://github.com/home-assistant/core/issues/86946 
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
I think should only be implemented for option flows, and not for all other flows?

**Comment by eifinger** [comment]
Yes. No scenario comes to my mind where this would be needed in a normal flow

**Comment by bramkragten** [comment]
As the code you have changed is generic for all flows, not just the config flow

**Comment by eifinger** [comment]
I have not yet contributed to the frontend, as far as I can see there is not yet a specific implementation for optionflows, am I correct?

**Comment by bramkragten** [comment]
The `flowConfig` (`this.flowConfig`) will be different for each type of flow, that means that this piece of code should be moved there. I also think the `undefined` check will need to be removed, as that could also be used to clear a field, but that will be filtered out in the JSON probably...

**Comment by eifinger** [comment]
I think the filter for empty/undefined can be moved to [config_flow.ts](https://github.com/home-assistant/frontend/blob/dev/src/data/config_flow.ts). This way it would not be used for config flows.

But do we need it in the first place? Why do we have to filter out the values for any flow at all?
Required fields are already checked before [here](https://github.com/home-assistant/frontend/blob/dev/src/dialogs/config-flow/step-flow-form.ts#L107)

**Comment by eifinger** [comment]
The change was made in https://github.com/home-assistant/frontend/pull/1662 to fix https://github.com/home-assistant/frontend/issues/1472

**Comment by eifinger** [comment]
I verified your assumption through local testing.
When setting the text field to empty the data being send to the backend includes the field value as `undefined` and when transformed to JSON it is not included.

This means that https://github.com/home-assistant/core/blob/dev/homeassistant/data_entry_flow.py#L245 stills sets the old default value. I currently think this has to be solved in the backend.

**Comment by eifinger** [comment]
Given that the issue is with default values for optional fields in the backend schema validation we somehow have to know whether the field was indeed not set or the user emptied id.

Through trial and error I determined that the field will have its "default" value when the user does not do anything and will be `undefined` when the user emptied a string field.

@bramkragten thank you for your previous input.
Ignoring the fact that the changes I made might/should be moved to OptionFlows only, can you think of any integrations where this might break existing functionality?

**Comment by bramkragten** [comment]
I honestly don't know, this is a problem we ran into a lot of times before already, always struggling with optional fields, clearing fields and partial updates...

**Comment by eifinger** [comment]
Setting it to `null` instead of the empty string would violate the `cv.string` schema:
```
User input malformed: string value is None for dictionary value @ data['incl_filter']
```

**Comment by eifinger** [comment]
Setting it to an empty string is also not an option because this would lead to problems with Optional fields which are not string, like int

**Comment by bramkragten** [comment]
> Setting it to `null` instead of the empty string would violate the `cv.string` schema:
> 
> ```
> User input malformed: string value is None for dictionary value @ data['incl_filter']
> ```

I think that is something that needs to be fixed in core, I think it is the only way for us to communicate that a field was cleared...

**Comment by eifinger** [comment]
I opened a PR in core to allow handling of `null`/`None` in config flow schemas

**Comment by eifinger** [comment]
Waiting on https://github.com/home-assistant-libs/voluptuous-serialize/pull/67

**Comment by eifinger** [comment]
Now that we have the info whether the field has `allow_none` we can set the values to `null` if the field was emptied.

Because the `data_schema` is not available in the config_flow and option_flow implementations I moved the logic back to `step-flow-form`

File: src/dialogs/config-flow/step-flow-form.ts [hunk 1]
```diff
@@ -161,6 +151,30 @@ class StepFlowForm extends LitElement {
...
+   */
+  private _transformEmptiedFields<T>(
+    data: Record<string, T>
+  ): Record<string, T | null> {
+    const toSendData: Record<string, T | null> = data;
+
+    this.step.data_schema
+      .filter((field) => field.allow_none === true)
+      .forEach((field) => {
+        if (toSendData[field.name] === undefined) {
```
Comment (bramkragten) [inline src/dialogs/config-flow/step-flow-form.ts:170]: We no longer handle empty strings now, should we add those too?

Comment (eifinger) [reply src/dialogs/config-flow/step-flow-form.ts:170]: If the field is an empty string it was sent by the backend as the default value and not changed by the user. As soon as the user enters characters and removes them again it becomes `undefined` if the user does not touch the field it is an empty string.

At least thats the behaviour I saw when testing the feature.

**Comment by epenet** [comment]
Note: one way to solve it in the backend is to set default to empty string on the schema, and use "suggested_value" for pre-populating the previous input.

**Comment by epenet** [comment]
I'm not sure this needs to be closed.
Even if it is not needed for https://github.com/home-assistant/core/issues/86946 it might still be needed in other scenarios.

For my part I am still struggling with https://github.com/home-assistant/core/pull/90412

**Comment by epenet** [comment]
Note: https://github.com/home-assistant/frontend/pull/14899 is somehow linked also

**Pull request closed without merging**